sklearn/metrics/_plot/confusion_matrix.py

```diff
@@ -261,7 +262,20 @@ def plot_confusion_matrix(estimator, X, y_true, *, labels=None,
 
     if display_labels is None:
         if labels is None:
-            display_labels = estimator.classes_
+            try:
+                display_labels = unique_labels(y_true, y_pred)
+            except TypeError as e:
+                # We expect y_true and y_pred to be of the same data type.
+                # If `y_true` was provided to the classifier as strings,
+                # `y_pred` given by the classifier will also be encoded with
+                # strings. So we raise a meaningful error
+                raise TypeError(
+                    f"Labels in y_true and y_pred should be of the same type. "
+                    f"Got y_true={np.unique(y_true)} and "
+                    f"y_pred={np.unique(y_pred)}. Make sure that the "
+                    f"predictions provided by the classifier coincides with "
+                    f"the true labels."
```

Comment:
Do we have a test to make sure this error is raised?

Reply:
I have removed the Try-Except wrapping as the function _confusion_matrix_(), which is used above to get the matrix itself, contains the same _unique_labels_() call that was wrapped by the Try-Except block, and the function _unique_labels_() raises an exception with a description when the true and predicted labels have different types. So if the execution arrives at the line, it will not make any problems.